prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package interrupts exposes helpers for graceful handling of interrupt signals
+package interrupts
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// we want to ensure that all registered servers and workers get a
+// change to gracefully shut down
+var wg = sync.WaitGroup{}
+
+// WaitForGracefulShutdown waits until all registered servers and workers
+// have had time to gracefully shut down. This function is blocking.
+func WaitForGracefulShutdown() {
+	wg.Wait()
+}
+
+// RunOrShutdown runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is blocking. Callers are
+// expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func RunOrShutdown(server *http.Server, gracePeriod time.Duration, logger *logrus.Logger) {
+	go Shutdown(server, gracePeriod, logger)
+	logger.WithError(server.ListenAndServe()).Info("Server exited.")
+}
+
+// Shutdown will shut down the server when an interrupt is received.
+// This function is blocking. Callers are expected to exit only after
+// WaitForGracefulShutdown returns to ensure all servers have had time
+// to shut down.
+func Shutdown(server *http.Server, gracePeriod time.Duration, logger *logrus.Logger) {
+	wg.Add(1)
+	defer wg.Done()
```

Comment:
In the current impl if I remove the `WaitForGracefulShutdown` function and `defer` a `wg.Wait()` before deferring the `wg.Done` it would make it impossible for any waiter to exit before they all do, but it won't deadlock.